### DISCUSSION TITLE
Why are there multiple URLs?

### DISCUSSION THREAD

**Discussion by ashhassan100** [digitalpetri/opc-ua-demo-server#7] [Q&A]
Hi. I downloaded one of the releases for the demo server and I ran the server on my mac. I noticed there are multiple URLs present in the terminal and I was wondering what each of them do. These are **opc.tcp://localhost:62541**, **opc.tcp://localhost:62541/milo**, **opc.tcp://localhost:4840/UADiscovery**. Some of them seem to connect and some don't. I was wondering what the difference is between these 3 URLs.

**Comment by kevinherron** [comment]
There's only two significant URLs that appear in the logs (excluding variations on hostnames):
- `opc.tcp://localhost:62541/milo`, the session endpoint
- `opc.tcp://localhost:62541/milo/discovery`, the discovery endpoint

The discovery endpoint has only discovery services available and can always be used without security. The session endpoint also offers discovery services but may require security depending on the configured security policies.

You probably saw `opc.tcp://localhost:4840/UADiscovery` in the config file, which is the endpoint URL of a discovery server to register with, if enabled.

**Reply by ashhassan100** [reply]
When I use ProSys OPC-UA Browser and type in **opc.tcp://localhost:62541/milo** as the connection address and **opc.tcp://localhost:62541/milo/discovery**, they both appear to just connect to the server. Isn't **opc.tcp://localhost:62541/milo/discovery** to list out the information about the server instead of connecting to the server? Also, I have noticed that all the endpoints bind to **0.0.0.0:62541**. What's the reason for this? And when I enter this into the browser, it says "No content in table."

**Reply by kevinherron** [reply]
>  Isn't opc.tcp://localhost:62541/milo/discovery to list out the information about the server instead of connecting to the server?

The endpoint URLs in the endpoints returned by the GetEndpoints service at the discovery endpoint are `opc.tcp://localhost:62541/milo`, and the Prosys client is following it as a correctly implemented client should.

> Also, I have noticed that all the endpoints bind to 0.0.0.0:62541. What's the reason for this? And when I enter this into the browser, it says "No content in table."

`0.0.0.0` is a meta-address that generally means "any IPv4 address". It's used instead of `localhost` or specifying the IP of one of the network adapters in your computer. You can change it if you want, but it's a safe default that allows connections from clients on other computers instead of just localhost.

**Reply by ashhassan100** [reply]
> The endpoint URLs in the endpoints returned by the GetEndpoints service at the discovery endpoint are opc.tcp://localhost:62541/milo, and the Prosys client is following it as a correctly implemented client should.

Does this mean that when a client connects to **opc.tcp://localhost:62541/milo/discovery** for an LDS, it won't directly connect to the server right? In this demo server, it connects directly to the server using GetEndpoints, but for an LDS,  that URL should list out all the servers registered to the LDS right?

**Reply by kevinherron** [reply]
If a client connects to an LDS, it's connecting to an LDS, yes. Not the server(s) registered with the LDS. The information you get from an LDS would tell you where to go next.

**Reply by ashhassan100** [reply]
I am just confused on the endpoints overall. I have read the specification by OPC and there are many overlaps and synonymous they use for endpoints. This is my understanding so far: when the demo server is live, it has a session endpoint and a discovery endpoint. Once the demo server is live, the ProSys Browser can connect to the discovery endpoint **opc.tcp://localhost:62541/milo/discovery** which in turn calls FindServer automatically and then getEndpoints and then establishes a connection with the server. This is all done automatically by ProSys browser and all we see is that the Browser connects to the server. We do not see the background processes of FindServers and getEndpoints?

**Reply by kevinherron** [reply]
When Prosys client asks you to select security settings what is available is based on the results of the GetEndpoints call.

I suspect if the Prosys client called FindServers and whatever it was connected to returned multiple server choices you would see a dialog for that as well.
